bindings/python/fdb/impl.py

```diff
@@ -593,7 +596,20 @@ def is_ready(self):
         return bool(self.capi.fdb_future_is_ready(self.fpointer))
 
     def block_until_ready(self):
-        self.capi.fdb_future_block_until_ready(self.fpointer)
+        # Checking readiness is faster than using the callback, so it saves us time if we are already
+        # ready. It also doesn't add much to the cost of this function
+        if not self.is_ready():
+            # Blocking in the native client from the main thread prevents Python from handling signals.
+            # To avoid that behavior, we implement the blocking in Python using semaphores and on_ready.
+            # Using a Semaphore is faster than an Event, and we create only one per thread to avoid the 
+            # cost of creating one every time.
+            semaphore = getattr(_thread_local_storage, 'future_block_semaphore', None)
+            if semaphore is None:
+                semaphore = multiprocessing.Semaphore(0)
+                _thread_local_storage.future_block_semaphore = semaphore
+
+            self.on_ready(lambda self: semaphore.release())
+            semaphore.acquire()
```

Comment:
I guess I can think of the following dead-lock possibility:

* Someone creates a future that ends up never completing (let's say they ask for a read version from a cluster that is down/not able to be connected to) and calls `block_until_ready()` so the semaphore has a value of -1
* The user gets tired of waiting and interrupts
* The user handles the signal (let's say they are running in a REPL and Ctrl-C brings them back to the shell)
* They begin a new future and call `block_until_ready()` (all on the same thread) so the semaphore has a value of -2
* The second future completes releasing the semaphore so it has a value of -1
* The user is still blocked as they were never able to acquire the semaphore

Does that sound right? And does that sound like something we need to worry about?

Reply:
Hmm, that seems like a valid concern

Reply:
I added a bit to replace the semaphore with a new one if an exception is raised during acquire. One way to produce this type of problem is to open up a REPL and try to connect to two clusters, one that's up and one that's not. 

Interestingly, the behavior I got when trying this wasn't exactly as described above. Instead, sometimes it worked, and sometimes it failed with a future_not_ready exception. The new version seems to avoid that.